File: TPERM2/TPERM2.ml
let rep n f =
        for _ = 1 to n do
                f ()
        done

let nth_char n =
        Char.chr (Char.code 'a' + n)

type dir = L | R

type elem = {
        ch: char;
        mutable d: dir;
}

let nth_elem n = {
        ch = nth_char n;
        d = L;
}

let gen_arr k =
        Array.init k nth_elem

let fold_arrayi fn arr =
        let len = Array.length arr in
        if len = 0 then None
        else
                let rec proc res i =
                        if i < len then
                                let next = (i, arr.(i)) in
                                proc (fn res next) (i + 1)
                        else Some res
                in
                let first = (0, arr.(0)) in
                proc first 1

let option_filter pred = function
        | None -> None
        | Some x as opt ->
                if pred x then opt
                else None

let get_elem_to_move arr =
        let len = Array.length arr in
        let nonfirst idx = idx > 0 in
        let nonlast idx = idx < len - 1 in
        let smaller_left idx = arr.(idx).ch > arr.(idx - 1).ch in
        let smaller_right idx = arr.(idx).ch > arr.(idx + 1).ch in
        let moveable (idx, el) =
                match el.d with
                | L -> nonfirst idx && smaller_left idx
                | R -> nonlast idx && smaller_right idx
        in
        let max_el ((_, e1) as it1) ((_, e2) as it2) =
                if (e1.ch > e2.ch && moveable it1) || not (moveable it2) then it1
                else it2
        in
        fold_arrayi max_el arr |> option_filter moveable

let swap arr id1 id2 =
        let tmp = arr.(id1) in
        arr.(id1) <- arr.(id2);
        arr.(id2) <- tmp;
        arr

let move_elem arr (idx, el) =
        match el.d with
        | L -> swap arr idx (idx - 1)
        | R -> swap arr idx (idx + 1)

let rev = function
        | L -> R
        | R -> L

let switch_dirs arr {ch; d} =
        Array.iter (fun el ->
                if el.ch > ch then el.d <- rev el.d
        ) arr;
        arr

let perms arr f =
        f arr;
        let rec proc = function
                | None -> ()
                | Some (idx, el as it) ->
                        let _ = move_elem arr it in
                        let _ = switch_dirs arr el in
                        let () = f arr in
                        let it' = get_elem_to_move arr in
                        proc it'
        in proc (get_elem_to_move arr)

let print_elem out_ch el =
        output_char out_ch el.ch

let print_arr out_ch arr =
        Array.iter (print_elem out_ch) arr

let print_perm perm =
        Printf.printf "%a\n" print_arr perm

let _ =
        let () =
                assert (nth_char 0 = 'a');
                assert (nth_char 1 = 'b');
                assert (nth_char 2 = 'c');
                assert (gen_arr 1 = [| {ch = 'a'; d = L} |]);
                assert (gen_arr 2 = [| {ch = 'a'; d = L}; {ch = 'b'; d = L} |]);
                assert (gen_arr 3 = [| {ch = 'a'; d = L}; {ch = 'b'; d = L}; {ch = 'c'; d = L};  |]);
                assert (get_elem_to_move [| |] = None);
                assert (get_elem_to_move [| {ch = 'a'; d = L} |] = None);
                assert (get_elem_to_move [| {ch = 'a'; d = L}; {ch = 'b'; d = L} |] = Some (1, {ch = 'b'; d = L}));
                assert (get_elem_to_move [| {ch = 'a'; d = L}; {ch = 'b'; d = R} |] = None);
                assert (get_elem_to_move [| {ch = 'b'; d = L}; {ch = 'a'; d = R} |] = None);
                assert (get_elem_to_move [| {ch = 'b'; d = L}; {ch = 'c'; d = L}; {ch = 'a'; d = L} |] = Some (1, {ch = 'c'; d = L}));
                assert (get_elem_to_move [| {ch = 'c'; d = L}; {ch = 'b'; d = R}; {ch = 'a'; d = L} |] = Some (1, {ch = 'b'; d = R}));
                assert (get_elem_to_move [| {ch = 'a'; d = L}; {ch = 'b'; d = L}; {ch = 'c'; d = R} |] = Some (1, {ch = 'b'; d = L}));
                assert (get_elem_to_move [| {ch = 'b'; d = L}; {ch = 'a'; d = L}; {ch = 'c'; d = R} |] = None);
                assert (move_elem [| {ch = 'a'; d = R}; {ch = 'b'; d = L} |] (0, {ch = 'a'; d = R}) = [| {ch = 'b'; d = L}; {ch = 'a'; d = R} |] );
                assert (switch_dirs [| {ch = 'a'; d = L}; {ch = 'b'; d = L}; {ch = 'c'; d = L} |] {ch = 'b'; d = L} = [| {ch = 'a'; d = L}; {ch = 'b'; d = L}; {ch = 'c'; d = R} |]);
                assert (switch_dirs [| {ch = 'b'; d = L}; {ch = 'c'; d = R}; {ch = 'a'; d = L} |] {ch = 'c'; d = R} = [| {ch = 'b'; d = L}; {ch = 'c'; d = R}; {ch = 'a'; d = L} |]);
        in
        Scanf.scanf "%d\n" (fun t ->
                rep t (fun () ->
                        Scanf.scanf "%d\n" (fun k ->
                                let arr = gen_arr k in
                                perms arr print_perm
                        )
                )
        )

File: TPERM2/TPERM2.lua
local L = 'L'
local R = 'R'

local function rep(n, f)
	for _= 1, n do
		f()
	end
end

local function gen_arr(k)
	local arr = {}
	for i = 1, k do
		arr[i] = {
			ch = string.char(string.byte('a') + i - 1),
			dir = L,
		}
	end
	return arr
end

local function get_elem_to_move(arr)
	local len = #arr
	local nonfirst = function(idx) return idx > 1 end
	local nonlast = function(idx) return idx < len end
	local smaller_left = function(idx) return arr[idx].ch > arr[idx - 1].ch end
	local smaller_right = function(idx) return arr[idx].ch > arr[idx + 1].ch end
	local moveable = function(idx, elem)
		local dir = elem.dir
		if dir == L then
			return nonfirst(idx) and smaller_left(idx)
		elseif dir == R then
			return nonlast(idx) and smaller_right(idx)
		else
			error('invalid direction ' .. dir)
		end
	end
	local function bigger_elem(i, j)
		local el_i = arr[i]
		local el_j = arr[j]
		if (el_i.ch > el_j.ch and moveable(i, el_i)) or not moveable(j, el_j) then
			return i
		else
			return j
		end
	end
	local function get_biggest_elem_to_move()
		local current = 1
		for i = 2, len do
			current = bigger_elem(current, i)
		end
		return current, arr[current]
	end
	local idx, elem = get_biggest_elem_to_move()
	if not moveable(idx, elem) then
		return false, {}
	end
	return idx, elem
end

local function swap(arr, i, j)
	arr[i], arr[j] = arr[j], arr[i]
end

local function move_elem(arr, idx, elem)
	local dir = elem.dir
	if dir == L then
		swap(arr, idx, idx - 1)
	elseif dir == R then
		swap(arr, idx, idx + 1)
	else
		error('invalid direction ' .. dir)
	end
end

local function reverse(elem)
	if elem.dir == L then
		elem.dir = R
	else
		elem.dir = L
	end
end

local function switch_dirs(arr, elem)
	for i = 1, #arr do
		local el = arr[i]
		if el.ch > elem.ch then
			reverse(el)
		end
	end
end

local function perms(arr, fn)
	fn(arr)
	while true do
		local idx, elem = get_elem_to_move(arr)
		if (idx == false) then
			break
		end
		move_elem(arr, idx, elem)
		switch_dirs(arr, elem)
		fn(arr)
	end
end

local function print_perm(arr)
	for i = 1, #arr do
		io.write(arr[i].ch)
	end
	io.write('\n')
end

local n = io.read()
rep(n, function()
	local k = io.read()
	local arr = gen_arr(k)
	perms(arr, print_perm)
end)
